File: day4/day4.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
)

type position struct {
	x int
	y int
}

const GRID_SIZE = 10

func main() {
	cwd, err := os.Getwd()
	if err != nil {
		fmt.Println(err)
	}

	data_path := filepath.Join(cwd, "day4", "data.txt")

	file, err := os.Open(data_path)
	if err != nil {
		fmt.Println(err)
	}
	fmt.Println(file)
	part_one(file)
}

func part_one(file *os.File) {
	scanner := bufio.NewScanner(file)
	var grid [GRID_SIZE][GRID_SIZE]rune
	var y = 0
	var x_pos []position
	var xmax_count = 0
	for scanner.Scan() {
		var line string = scanner.Text()
		for x, char := range line {
			grid[y][x] = char
			if char == 'X' {
				x_pos = append(x_pos, position{x, y})
				fmt.Print(string(char))
			} else {
				fmt.Print(".")
			}

		}
		fmt.Println("")
		y++
	}
	for _, pos := range x_pos {
		if check_xmas(pos, grid, 0) {
			xmax_count++
		}
	}
	fmt.Println(len(x_pos))
	fmt.Println(xmax_count)

}

func check_xmas(start position, grid [GRID_SIZE][GRID_SIZE]rune, index int) bool {
	if start.x == -1 && start.y == -1 {
		return false
	}
	var i = index
	var xmas = []rune{'M', 'A', 'S'}
	directions := []string{"U", "D", "L", "R", "UL", "UR", "DL", "DR"}

	for _, dir := range directions {
		var pos = check(start, dir, xmas[i], grid)
		if i+1 == len(xmas) || check_xmas(pos, grid, i+1) {
			return true
		}
	}
	return false
}
func check(curr position, direction string, letter rune, grid [GRID_SIZE][GRID_SIZE]rune) position {
	var pos position

	switch direction {
	case "U":
		pos = position{curr.x, curr.y + 1}
	case "D":
		pos = position{curr.x, curr.y - 1}
	case "L":
		pos = position{curr.x + 1, curr.y}
	case "R":
		pos = position{curr.x - 1, curr.y}
	case "UL":
		pos = position{curr.x + 1, curr.y + 1}
	case "UR":
		pos = position{curr.x - 1, curr.y + 1}
	case "DL":
		pos = position{curr.x - 1, curr.y - 1}
	case "DR":
		pos = position{curr.x + 1, curr.y - 1}
	}

	if get_from_grid(pos, grid) == letter {
		return pos
	}

	return position{-1, -1}
}

func get_from_grid(pos position, grid [GRID_SIZE][GRID_SIZE]rune) rune {
	if 0 <= pos.x && pos.x < len(grid[0]) && 0 <= pos.y && pos.y < len(grid) {
		return grid[pos.y][pos.x]
	}
	return ' '
}

File: day4/day4.py
import os

DATA_PATH = os.path.join(os.getcwd(), "day4", "data.txt")


def check_grid(i, j, grid):
    if 0 <= i < len(grid) and 0 <= j < len(grid[0]):
        return grid[i][j]
    # print("invalid", i, j)
    return ""


def check_xmas(i, j, grid, xi=0, dir=(0, 0)):
    # print(dir) if dir != (0, 0) else 0
    count = 0
    xmas = "XMAS"
    dirs = [(1, 1), (1, 0), (0, 1), (-1, 0), (0, -1), (-1, -1), (-1, 1), (1, -1)]
    if xi == 0:
        # print("START", i, j)
        for dir in dirs:
            if xmas[xi] == check_grid(i, j, grid):
                count += check_xmas(i + dir[0], j + dir[1], grid, xi + 1, dir)
        return count
    else:
        if xmas[xi] == check_grid(i, j, grid):
            if xi + 1 == len(xmas):
                # print("XMAS!!!")
                return 1
            return check_xmas(i + dir[0], j + dir[1], grid, xi + 1, dir)
        else:
            # print(f"{xmas[xi]} == {check_grid(i, j, grid)}", "FAILED")
            return 0


if "__main__" == __name__:
    grid = []
    with open(DATA_PATH) as lines:
        for line in lines:
            row = [char for char in line if not char == "\n"]
            grid.append(row)

    xmas_count = 0
    x_loc = []
    xmas_loc = []
    for i, row in enumerate(grid):
        for j, char in enumerate(row):
            if char == "X":
                x_loc.append((i, j))
                before = xmas_count
                xmas_count += check_xmas(i, j, grid)
                if before < xmas_count:
                    xmas_loc.append((i, j))

    print(xmas_count)
    # print(xmas_loc)
    # print(x_loc)
